lockfile/src/parse_depfile.rs

```diff
@@ -0,0 +1,250 @@
+
+    // Attempt to parse the identified lockfile.
+    let mut lockfile_error = None;
+    if maybe_lockfile {
+        // Parse lockfile content.
+        let packages = parse_lockfile_content(contents, parser);
+
+        match packages {
+            Ok(packages) => return Ok(ParsedLockfile::new(id, format, packages)),
+            // Store error on failure.
+            Err(err) => lockfile_error = Some(err),
+        }
+    }
+
+    // Check if file could be a manifest based on file name.
+    let maybe_manifest = pseudopath.as_ref().map_or(false, |path| parser.is_path_manifest(path));
+
+    // Generate lockfile if path might be a manifest and feature and option are
+    // enabled.
+    #[cfg(feature = "generator")]
+    if let Some(generation_path) = _generation_path.filter(|_| maybe_manifest) {
+        return Ok(generate_lockfile(&generation_path, id, format, parser)?);
+    }
+
+    // Return the original lockfile parsing error.
+    match lockfile_error {
+        // Report parsing errors only for lockfiles.
+        Some(err) if !maybe_manifest => Err(err),
+        _ => Err(ParseError::ManifestWithoutGeneration(id.into())),
+    }
+}
+
+/// Attempt to get packages from an unknown lockfile type
+fn try_get_packages(id: &str, contents: &str) -> Result<ParsedLockfile, anyhow::Error> {
+    for format in LockfileFormat::iter() {
+        let parser = format.parser();
+        if let Some(packages) = parser.parse(contents).ok().filter(|pkgs| !pkgs.is_empty()) {
+            log::info!("Identified lockfile type: {}", format);
+
+            let packages = filter_packages(packages);
+
+            return Ok(ParsedLockfile::new(id, format, packages));
+        }
+    }
+
+    Err(anyhow!("Failed to identify type for lockfile {id:?}"))
+}
+
+/// Generate a lockfile from a manifest path.
+#[cfg(feature = "generator")]
+fn generate_lockfile(
+    path: &Path,
+    id: &str,
+    format: LockfileFormat,
+    parser: &dyn Parse,
+) -> Result<ParsedLockfile, anyhow::Error> {
+    // Find the generator for this format.
+    let generator = match parser.generator() {
+        Some(generator) => generator,
+        None => return Err(anyhow!("unsupported manifest file {id:?}")),
+    };
+
+    eprintln!("Generating lockfile for manifest {id:?} using {format:?}…");
+
+    // Generate a new lockfile.
+    let canonical_path = path.canonicalize()?;
+    let generated_lockfile = generator
+            .generate_lockfile(&canonical_path)
+            .context("Lockfile generation failed! For details, see: \
+                https://docs.phylum.io/docs/lockfile_generation")?;
+
+    // Parse the generated lockfile.
+    let packages = parse_lockfile_content(&generated_lockfile, parser)?;
+
+    Ok(ParsedLockfile::new(id, format, packages))
+}
+
+/// Attempt to parse a lockfile.
+fn parse_lockfile_content(
+    content: &str,
+    parser: &dyn Parse,
+) -> Result<Vec<PackageDescriptor>, ParseError> {
```

Comment:
If we remove `try_get_packages()`, I think this function would be a better entry point than `parse_depfile()` for applications that don't support lockfile generation.

The `parser` argument could be changed to `format: LockfileFormat` for consistency. If the caller doesn't know the lockfile format, they can use `get_path_format()`, which is already public.

The cases where `get_path_format()` doesn't work (either because no format matches or because the caller doesn't have a file path) might be worth further discussion... But I think `try_get_packages()` is too brittle to be a good answer there.